### ISSUE TITLE
support for units with pint

### ISSUE THREAD

**Issue by keewis** [pydata/xarray#3594]
`pint`'s implementation of NEP-18 (see hgrecco/pint#905) is close enough so we can finally start working on the `pint` support (i.e. make the integration tests pass).

Failing tests:
* top level functions
  - [ ] `align`: needs more investigation
  - [ ] `combine_by_coords`: needs more investigation
  - [ ] `combine_nested`: needs more investigation
  - [ ] `concat`: needs more investigation
  - [ ] `merge`: needs more investigation
  - [ ] `full_like`: `pint` currently does not implement `copyto`
  - [ ] `where`: needs more investigation
  - [ ] `dot`: `pint` currently does not implement `einsum`
* DataArray methods
  - [ ] `__init__`: indexes strip units, blocked by #1603?
  - [ ] aggregation: `pint` currently does not implement `all`, `any`, `median`, `prod`, `sum`. `cumprod` works but is not allowed for quantities with dimensions (update the test?)
  - [ ] binary operations: matrix multiply (not implemented by `pint`, yet: hgrecco/pint#865)
  - [ ] bivariate ufuncs: `pint` does not implement `maximum`
  - [ ] numpy methods: `pint` does not implement `rank`
  - [ ] `ffill`, `bfill`: needs more investigation
  - [ ] `fillna`: calls `astype` on fill value which may be a python scalar wrapped by `pint.Quantity` and needs more investigation
  - [ ] `isin`: `pint` does not implement `isin`
  - [ ] `where`: bug in the test code and needs more investigation
  - [ ] `interpolate_na`: needs more investigation
  - [ ] `combine_first`: needs more investigation
  - [ ] `equals`, `identical`: needs more investigation
  - [ ] `broadcast_like`: needs more investigation
  - [ ] `drop_sel`: test bug / wrong method (use `drop_vars`)
  - [ ] `astype`: needs more investigation
  - [ ] `item`: needs more investigation
  - [ ] `drop_sel`: needs more investigation
  - [ ] `sel`, `loc`: indexes strip units (#1603)
  - [ ] `squeeze`: needs more investigation
  - [ ] `head`, `tail`, `thin`: quantity not accepted as indexer
  - [ ] `interp_like`: needs more investigation
  - [ ] `reindex_like`: needs more investigation
  - [ ] `to_unstacked_dataset`: needs more investigation
  - [ ] `shift`, `roll`: needs more investigation
  - [ ] `quantile`, `dot`: `pint` does not implement `nanpercentile` and `einsum`
  - [ ] `groupby`: test bug?
  - [ ] `groupby_bins`, `rolling`: needs more investigation
  - [ ] `rolling_exp`: numbagg
  - [ ] `resample`: needs more investigation
  - [ ] `assign_coords`: needs more investigation
* Dataset methods
  - [ ] `__init__`: indexes strip units and merge does not work yet (test bug is also possible)
  - [ ]  aggregation: `pint` does not implement `all`, `any`, `xarray` does not implement `np.median` and `cumprod` is not allowed for quantities with dimensions
  - [ ] `rank`: `pint` does not implement `rank`
  - [ ] `ffill`, `bfill`: needs more investigation
  - [ ] `fillna`: needs more investigation
  - [ ] `isin`: `pint` does not implement `isin`
  - [ ] `where`: needs more investigation
  - [ ] `interpolate_na`: needs more investigation
  - [ ] `combine_first`: needs more investigation
  - [ ] `equals`, `identical`: needs more investigation
  - [ ] `broadcast_like`: needs more investigation
  - [ ] `to_stacked_array`: needs more investigation
  - [ ] `shift`, `roll`: needs more investigation
  - [ ] `isel`: needs more investigation
  - [ ] `sel`, `loc`: indexes strip units
  - [ ] `head`, `tail`, `thin`: needs more investigation
  - [ ] `interp`, `reindex`: needs more investigation
  - [ ] `interp_like`, `reindex_like`: needs more investigation
  - [ ] `quantile`: `pint` does not implement `nanpercentile`
  - [ ] `map`: `pint` does not implement `fabs` (but we could probably use a different ufunc here)
  - [ ] `groupby`, `groupby_bins`, `rolling`, `rolling_exp`: needs more investigation
  - [ ] `resample`: needs more investigation
  - [ ] grouped operations: `assign`, `assign_coords`: needs more investigation (probably same issue as `groupby`)
  - [ ] `drop_sel`: test bug, use `drop_vars` here
  - [ ] `merge`: needs more investigation

I will try to figure out the reason for each of these test failures, but I'd appreciate help.

**Comment by dcherian** [comment]
Thanks for leading this effort @keewis.

I would start with the lowest-level operations like the constructors and `align`, `concat`, `merge`. These are called in many of the other functions so fixing this is a prerequisite for getting the rest working. I've looked at align, concat & merge recently so can help if you need to chat about confusing error messages. 

> indexes strip units

What does this mean? You can't have units in a IndexVariable?

**Comment by keewis** [comment]
> What does this mean? You can't have units in a IndexVariable?

yes, we had that discussion from https://github.com/pydata/xarray/issues/525#issuecomment-514452182 on. Short version: `pd.Index` converts using `np.asarray` and support for units probably requires #1603.

**Comment by amcnicho** [comment]
> I will try to figure out the reason for each of these test failures, but I'd appreciate help.

Would #3643 be the best place to offer contributions at this point, or somewhere else?

**Comment by keewis** [comment]
I think issues related to `DataArray` + `pint` should be in #3643, for everything else you can use this issue or new issues / pull requests.

If you want to, I'd appreciate someone reviewing the tests in `test_units.py` since I don't think anyone other than me thoroughly looked at them. You could also investigate / fix the `Dataset` issues, investigate the reason for the `UnitStrippedWarnings` or start writing documentation on how to use `pint` in combination with `xarray`.

**Comment by keewis** [comment]
so, except from the major issues mentioned above which we won't be able to fix in the near future (but there will probably be workarounds in `pint-xarray`) we only have three minor issues: `nanprod`, support for per variable fill values and the repr (#2773).

I don't think `nanprod` and the fill values are particularly urgent, so if we get support for the `repr` (maybe using some sort of hook that a library like `pint-xarray` can then use to properly format the duckarray) and put together the documentation page, we could include this in the `0.16` release.

**Comment by keewis** [comment]
#6873 might fix the `nanprod` issue, and we have a separate issue for the last big change left (#3950, which is not really limited to quantities) so I agree that we should be able to close this.

We might want to open a new issue to get the [known issues](https://docs.xarray.dev/en/latest/user-guide/duckarrays.html#missing-features) to work, though.

**Comment by tennlee** [comment]
I just came here from https://xarray.dev/blog/introducing-pint-xarray and wanted to leave a note of thanks for this work. Great job, much appreciated.
